Support tunnel routing in Multi-Pool IPAM mode

pippolo84 opened the pull request:
Rework of https://github.com/cilium/cilium/pull/38015 after the introduction of clusterID aware ipcache in https://github.com/cilium/cilium/pull/38379

The main difference with https://github.com/cilium/cilium/pull/38015 is commit https://github.com/cilium/cilium/pull/38483/commits/fa35fa0247909ada83f72952daef85a284bc1f40 where we take advantage of the `ClusterPrefix` parameter type in ipcache API to enrich the node prefixes (namely, the CiliumInternalIP, the Ingress and health IPs and the v4/v6 allocation CIDRs) with the result of a customizable PrefixClusterMutatorFn. This guarantees feature parity with the  tunnel map previously used for node allocation CIDRs.

Description of the previous version of the PR (https://github.com/cilium/cilium/pull/38015) below:

---

This PR introduces support for tunnel routing with the Multi-Pool IPAM mode. This is achieved by removing the reliance on the tunnel map as a fallback mechanism and instead use Pod CIDR entries in IPCache that act as this fallback. The pod CIDR entries contain the node's tunnel endpoint and encryption key and have the `reserved:world` identity (i.e. the same identity as the existing catch-all fallback, which would have been hit before this PR). 

The bulk of the work in this PR was done by @pippolo84. This PR is a replacement for his PR https://github.com/cilium/cilium/pull/37146 with a few notable changes:

1. The first commit has been reworked to add unit tests and fix up some issues in the previous version (c.f. https://github.com/cilium/cilium/pull/37146#discussion_r1973593637).
2. A commit has been added adding some unit tests to the IPCache package emulating some of the expected corner cases.
3. Commit `node: Remove insertions and deletions to tunnel map` has been dropped. In other words, even though the tunnel map is no longer used in the `main` branch, we still populate it. This addresses concerns with regards to downgrades.
4. The Conformance Multi-Pool workflow has been extended to also test in tunnel mode.

This change will eventually allow us to remove the tunnel map completely (ref https://github.com/cilium/cilium/issues/20170).

---

```release-note
Add support for tunnel routing in multi-pool IPAM mode
```

pippolo84 committed:
daemon: Remove checks for multi-pool IPAM and tunneling

As a consequence of using the ipcache map to support tunneling,
multi-pool IPAM is now compatible with it, thus lift the startup checks
that stops the agent when both the features are enabled.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node: Consider all node CIDRs when deleting a node

When handling the deletion of a node, we should take into account all
the node allocations CIDRs, not just the primary one.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node/manager: Store node allocation CIDRs into ipcache

Store allocation CIDRs from each remote node into the ipcache, using the
world identity.  This sets the ground for using the ipcache for
tunneling instead of relying on the additional tunnel map.

Co-authored-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
bpf: Use info from ipcache map lookup for encap

The ipcache map now stores entries describing remote node pod CIDRs,
including the tunnel endpoint and encryption key.

Therefore, when a packet must be encapsulated or encrypted we can
leverage the information from the lookup into the ipcache, avoiding the
additional one into the tunnel map.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro committed:
ipcache: Add unit tests for pod CIDR patterns

This commit adds a unit test asserting the expected behavior for pod
CIDR entries upserted to IPCache by the node manager.

In particular, we also test two corner cases:

  - The first case tests a scenario where the pod CIDR is also selected
    by a second resource, ensuring that tunnel and encryption key are
    preserved.

  - The second case tests a scenario where the pod CIDR is a /32 and the
    legacy API upserts the pod IP (e.g. based on CEP or kvstore),
    thereby shadowing the existing pod CIDR entry.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro committed:
ci/multi-pool: Add test for tunnel mode

As it is now supported.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
node/manager: Add configurable cluster mutator in Node obj handling

Introduce the possibility of configuring a prefix cluster mutator to
customize the prefixes from the Node object.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

giorio94 commented on pkg/node/manager/manager.go:
```diff
@@ -987,13 +1018,47 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+				worldLabelForPrefix(prefix),
+				ipcacheTypes.TunnelPeer{Addr: tunnelIP},
+			}
+			if m.nodeAddressHasEncryptKey() {
```
The EncryptKey should be always configured here. `nodeAddressHasEncryptKey` applies to node addresses only, where it should be set only if node-to-node encryption is enabled.

giorio94 commented on pkg/node/manager/manager.go:
```diff
@@ -877,6 +889,25 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+			mu = append(mu, entry)
+			podCIDRsAdded = append(podCIDRsAdded, entry.Prefix.AsPrefix())
+		}
+		m.ipcache.UpsertMetadataBatch(mu...)
```
This problem potentially also exists for the other async ipcache upsertions performed here (and this is likely not the best venue to fix it), and it lies in the fact that we are not waiting on the updates to be actually propagated down to the ipcache map. However, mentioning here as somewhat related, and that could theoretically lead to drops upon endpoint regeneration in case the ipcache is not fully populated yet (probably not caused by the missing fallback CIDRs though, as they should serve as a fallback only). 

AFAIU, we do wait on the first revision during daemon startup \[1]. Yet, while at that point we are guaranteed to have already synchronized with Kubernetes, no such guarantee exists for the kvstore and/or clustermesh (while we do block endpoint regeneration waiting on them a bit later), and a part of the initial node update events could be received afterwards as well. That said, simply waiting here would not work either (and go against the async principle), because the ipcache may not even have been started yet.

This is just to say that we should likely uniform the wait for sync process, and ensure that we wait for the first ipcache revision only after ensuring that all possible sources are synchronized. This is the more important the more subsystems we migrate to the asynchronous ipcache.

\[1]: https://github.com/cilium/cilium/blob/3a382254dab57eccd2bc5fb28de4537d7395d4be/daemon/cmd/daemon_main.go#L1694-L1696

pippolo84 replied:
Opened https://github.com/cilium/cilium/issues/38520 to discuss this further and clarify the possible impact.

giorio94 commented on daemon/cmd/daemon_main.go:
```diff
@@ -1330,10 +1330,6 @@ func initEnv(vp *viper.Viper) {
 
-	if option.Config.TunnelingEnabled() && option.Config.EnableAutoDirectRouting {
-		log.Fatalf("%s cannot be used with tunneling. Packets must be routed through the tunnel device.", option.EnableAutoDirectRoutingName)
-	}
```
This removal doesn't look related to me.

giorio94 commented on .github/workflows/conformance-multi-pool.yaml:
```diff
@@ -81,6 +81,15 @@ jobs:
           sha: ${{ inputs.SHA || github.sha }}
 
   multi-pool-ipam-conformance-test:
+    strategy:
```
Nit. Given that this is now a matrix, it could be worthy moving the "Wait for images" step outside. E.g., \[1]

\[1]: https://github.com/cilium/cilium/blob/ade6c0c6a0ff3a279d0002722bd6c0073bcc6d9c/.github/workflows/tests-e2e-upgrade.yaml#L81-L94

giorio94 replied:
The `wait-for-images` job should be added to the list of `needs` of `multi-pool-ipam-conformance-test`.

pippolo84 replied:
Updated, thanks for the heads up!

giorio94 commented on pkg/node/manager/manager.go:
```diff
@@ -818,10 +822,16 @@ func (m *manager) NodeUpdated(n nodeTypes.Node) {
+		prefix := ip.IPToNetPrefix(address.IP)
+		var prefixCluster cmtypes.PrefixCluster
+		if address.Type == addressing.NodeCiliumInternalIP {
+			prefixCluster = cmtypes.PrefixClusterFrom(prefix.Addr(), prefix.Bits(), m.prefixClusterMutatorFn(&n)...)
```
Nit. `cmtypes.PrefixClusterFrom` could be likely changed to directly take a prefix as parameter, rather than address and bits.

pippolo84 replied:
Nice idea :+1:

giorio94 reviewed (changes requested):
Thanks! A couple of comments inline, besides that looks great to me.

Artyop reviewed (approved):
besides Marco's review about waiting for image step, looks good for github-sec and ci-structure

pippolo84 committed:
ci/multi-pool: Move the wait for images step before the actual test

Doing that the workflow waits for the images just once before running
the test matrix.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
clustermesh: Change PrefixClusterFrom to accept a netip.Prefix

Since the vast majority of the callers already have a netip.Prefix
instance, improve PrefixClusterFrom helper so that it takes a
netip.Prefix instead of an address and the bit prefix length.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

giorio94 reviewed (approved):
One more comment inline (above), looks good to me otherwise.
